text/0000-natvis.md

```diff
@@ -0,0 +1,491 @@
+selects the `find_c_defines` call frame and looks at the local variables
+window.
+
+Unfortunately, the debugger's view of the `captures` variable does not give
+her any useful information at all. It shows only something like:
+
+```text
+> Variables
+  > captures: {...}
+    > text: "...the entire input text..."
+    > locs: {...}
+      > __0: (4) vec![None, None, None, None]
+      > named_groups: (refs:2) size=0, capacity=1
+        > [raw]: alloc::sync::Arc<std::collections::hash::map::HashMap<...>>
+          > ptr: {pointer:0xNNNNNNNN}
+            > pointer: {...}
+              > strong: {...}
+              > weak: {...}
+              > data: size=0, capacity=1
+                > base: {...}
+                  > hash_builder: {...}
+        ...
+```
+
+The debugger shows the structure of the data, not its meaning. It is not very
+useful for Carol. Even the implementor of `regex` would have a hard time knowing
+how to decode this. In reality, when trying to understand the state of the captures
+variable there are several methods defined for a `Captures` type that paint
+the actual picture in terms of the information a Rust developer would like to
+extract from this variable. In order to meaningfully understand what the `Captures`
+type is truly trying to tell us, it would be very helpful to visualize this data
+differently in the debugger.
+
+What we _want_ is something like this:
+
+```text
+> Variables:
+  > captures: {...}
+    > 1: "SOME_CONSTANT"
+    > 2: "42"
+    > 3: "// some developer comment"
+```
+
+This RFC will describe how to support adding Natvis visualizations which is supported by:
+
+* The Windows Debugger (WinDbg)
+* Visual Studio Debugger.
+
+It should be easy for Rust developers to add debugger visualizations to their
+crates.
+
+## Supporting Natvis
+
+This section describes how Microsoft's Natvis is supported in Rust.
+
+To use Natvis, developers write XML documents that describe how debugger types
+should be displayed using the natvis schema. (See: https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019)
+The Natvis files provide patterns, which match type names, and for matching
+types, a description of how to display those types. This allows for some
+limited support for generic types.
+
+Rust developers can add one or more `.natvis` files to their crate. The
+`Cargo.toml` file specifies the path of these Natvis files via a new
+manifest key. Cargo would then pass the set of `.natvis` files that were
+specified in `Cargo.toml` and pass it to the rustc invocation for the crate
+via a new `-C` flag and enabling unstable options in Rust via the
+`-Z unstable-options`. See the below example for how rustc would embed these
+Natvis files in the debuginfo for a binary as well as new compiler options
+to be added to rustc.
+
+To provide Natvis files, developers create a file with the `.natvis` file
```

Comment:
Since the name is specified in full in Cargo.toml, I would expect it to be unnecessary for the file to have the .natvis extension (merely convention, or for use for tools outside the Rust ecosystem)?

Reply:
`.natvis` is the conventional extension used by Windows dev tools (Visual Studio, etc.) to identify NatVis files. I think it is valuable to stick with this convention.

Reply:
WinDbg is unable to load a Natvis file if it was embedded into the PDB via the linker without the `.natvis` extension. This is the same as how the Rust compiler currently adds Natvis for std and other libs i.e., `libstd.natvis`.